Add Seccomp Notify support (#1)
kinvolk/runtime-spec: config.md

Hooks allow users to specify programs to run before or after various lifecycle events.
Hooks MUST be called in the listed order.
All hooks MUST be passed a data structure over stdin so that they may do work appropriately.

> Exect when specified otherwise above, the data structure is the [state](runtime.md#state) of the container.

Reviewer: ```suggestion
Except when specified otherwise above, the data structure is the [state](runtime.md#state) of the container.
```

The definition of `createContainer` hooks is currently underspecified and hooks authors, should only expect from the runtime that the mount namespace and different mounts will be setup. Other operations such as cgroups and SELinux/AppArmor labels might not have been performed by the runtime.

### <a name="configHooksSendSeccompFd" />SendSeccompFd Hooks

> The `sendSeccompFd` hooks MUST only be called if the seccomp policy contains `SCMP_ACT_NOTIFY`.

Reviewer: ```suggestion
The `sendSeccompFd` hooks MUST only be called if the seccomp policy contains `SCMP_ACT_NOTIFY` actions.
```

The `sendSeccompFd` hooks MAY additionally be called while the container is in the [`running` state](runtime.md#runtimeState), for example during an `exec` operation.
The goal of this hook is to pass the seccomp file descriptor to a seccomp agent.

The `sendSeccompFd` hooks' path MUST resolve in the [runtime namespace](glossary.md#runtime-namespace).
The `sendSeccompFd` hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).

### <a name="seccompstate" />The Seccomp State

The seccomp state is a data structure passed via stdin to the SendSeccompFd hooks.
It includes the following properties:

* **`ociVersion`** (string, REQUIRED) is version of the Open Container Initiative Runtime Specification with which the seccomp state complies.
* **`phase`** (string, REQUIRED) is the phase at which the seccomp file descriptor is created.
    The value MAY be one of:

    * `start`: the seccomp filter is created following the [`start`](runtime.md#start) command

>     * `exec`: the seccomp filter is created following an `exec` command

Reviewer: I just realized the OCI spec doesn't define an `exec` command. Given that, I'm not sure about including this field in the state anymore. 

Maybe we could use `State.Status` to differentiate between a call starting the container or a call when the container is already running?

Author: sounds good if it works. Have you checked what does `State.Status` contain when the seccomp hook is called from the startup phase?

Reviewer: Yes, it's "creating", hence the hook / agent is able to differentiate the two cases based on it.

Author: Checked, it is State.Status=`creating` and State.Status=`running`. Looks good.

Author: I removed the "phase" field.

The `poststop` hooks' path MUST resolve in the [runtime namespace](glossary.md#runtime-namespace).
The `poststop` hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).

> ### <a name="configHooksSendSeccompFd" />SendSeccompFd

Reviewer: What about adding it before the `StartContainer Hooks` to keep them in the same order they are called?

> The `sendSeccompFd` hooks is only called if the seccomp policy contains `SCMP_ACT_NOTIFY`.

Reviewer: ```suggestion
The `sendSeccompFd` hooks MUST be only called if the seccomp policy contains `SCMP_ACT_NOTIFY`.
```

The `sendSeccompFd` hooks MUST be called after the [`start`](runtime.md#start) operation is called and after the seccomp policy is installed but [before the user-specified program command is executed](runtime.md#lifecycle).
The goal of this hook is to pass the seccomp file descriptor to a seccomp agent.

The `sendSeccompFd` hooks' path MUST resolve in the [runtime namespace](glossary.md#runtime-namespace).

> The `peccompFdoststop` hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).

Reviewer: ```suggestion
The `sendSeccompFd` hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).
```

:laughing:

| `startContainer`        | container | After the start operation is called but before the user-specified program command is executed.                                     |
| `poststart`             | runtime   | After the user-specified process is executed but before the start operation returns.                                               |
| `poststop`              | runtime   | After the container is deleted but before the delete operation returns.                                                            |

> | `sendSeccompFd`         | runtime   | After the start operation is called but before the user-specified program command is executed.                                     |

Reviewer: Same here, what about putting this in the same order they are called?
